tests/codegen.rs

```diff
@@ -1,118 +1,71 @@
+use x509_parser::x509::X509Name;
 
 #[tokio::test]
-async fn generated_code_is_fresh() {
-    // Fetch the list of certificates as a PEM file from mkcert.org
+async fn new_generated_code_is_fresh() {
+    // Configure a Reqwest client that only trusts the CA certificate expected to be the
+    // root of trust for the CCADB server.
+    let root = include_bytes!("data/DigiCertGlobalRootCA.der");
+    let root = reqwest::Certificate::from_der(root).unwrap();
+    let client = reqwest::Client::builder()
+        .user_agent(format!("webpki-roots/v{}", env!("CARGO_PKG_VERSION")))
+        .add_root_certificate(root)
+        .build()
+        .unwrap();
+
+    let ccadb_url =
+        "https://ccadb-public.secure.force.com/mozilla/IncludedCACertificateReportPEMCSV";
+    eprintln!("fetching {ccadb_url}...");
+
+    let req = client.get(ccadb_url).build().unwrap();
+    let csv_data = client
+        .execute(req)
+        .await
+        .expect("failed to fetch CSV")
+        .text()
+        .await
+        .unwrap();
+
+    // Parse the CSV metadata.
+    let metadata = csv::ReaderBuilder::new()
+        .has_headers(true)
+        .from_reader(csv_data.as_bytes())
+        .into_deserialize::<CertificateMetadata>()
+        .collect::<Result<Vec<_>, _>>()
+        .unwrap();
+
+    // Filter for just roots with the TLS trust bit that are not distrusted as of today's date.
+    let mut trusted_tls_roots = metadata
+        .into_iter()
+        .filter(|root| root.trusted_for_tls(&Utc::now().naive_utc().date()))
+        .collect::<Vec<CertificateMetadata>>();
+
+    // Sort roots by FP for consistent ordering.
+    trusted_tls_roots.sort_by(|a, b| a.sha256_fingerprint.cmp(&b.sha256_fingerprint));
 
-    let mut except = String::with_capacity(128);
-    for (i, ca) in EXCLUDED_CAS.iter().enumerate() {
-        if i > 0 {
-            except.push('+');
-        }
-
-        except
-            .write_fmt(format_args!(
-                "{}",
-                percent_encoding::percent_encode(ca.as_bytes(), percent_encoding::NON_ALPHANUMERIC,)
-            ))
-            .unwrap();
-    }
-
-    let url = format!("https://mkcert.org/generate/all/except/{except}");
-    eprintln!("fetching {url}...");
-    let body = reqwest::get(&url).await.unwrap().text().await.unwrap();
-    fs::write("fetched.pem", &body).unwrap();
-
-    // Split file contents into a Vec<Vec<&str>> where each inner Vec represents one certificate
-
-    let mut certs = Vec::with_capacity(64);
-    let mut current = Vec::with_capacity(16);
-    for ln in body.lines() {
-        if !ln.is_empty() {
-            current.push(ln);
-            continue;
-        } else if current.is_empty() {
-            continue;
-        }
-
-        certs.push(current);
-        current = Vec::with_capacity(16);
-    }
-
-    certs.push(current);
-
-    // Parse each certificate and check fingerprints
-
-    let mut hashed = HashMap::<Vec<u8>, (Vec<&str>, Vec<u8>)>::default();
```

Comment:
This was maybe a bit unclear and not well ported from Python to Rust, but the Python version made a point of checking for duplicate certs:

https://github.com/rustls/webpki-roots/pull/34/files#diff-b70bfb027a8ef3ec4ab07a53a3b8ba83e63334ba3ce5ca6d9d773900ed452e5eL190

Maybe we should add that back?

Reply:
I've re-added the check for duplicate fingerprints.